Writer.prototype.parse to cache by tags in addition to template string

raymond-lam opened the pull request:
Resolve #617 
Because the results of `Writer.prototype.parse` depends on both the template string and the tags, it should cache by both

dasilvacontin commented on test/parse-test.js:
```diff
@@ -103,4 +103,36 @@ describe('Mustache.parse', function () {
+  describe('when parsing a template after already having parsed that template with a different Mustache.tags', function() {
+    it('returns different tokens for the latter parse', function() {
+      var template = "{{foo}}[bar]";
+      var parsedWithBraces = Mustache.parse(template, ['(', ')']);
```
These are parenthesis, not braces, right?

raymond-lam replied:
Yes, my bad. I'll fix it.

raymond-lam commented:
@dasilvacontin Addressed your comment -- good catch! I actually meant to test the template with `Mustache.tags` as the delimiters, with it being braces.

raymond-lam commented:
@dasilvacontin Awesome, thanks!

dasilvacontin commented:
I started refreshing my memory about the other PR, but I don't have enough time today. Will do tomorrow morning.

raymond-lam commented:
@dasilvacontin no rush! Thanks!

petrkoutnysw commented:
I am afraid this PR caused breaking change for us.
We are doing following:

1. parse our template with custom tags
2. render the template - according to mustache.js:465 our custom tags are not used, hence the tokens are different and custom tags are not recognized

phillipj commented:
@petrkoutnysw a bug fix landed after this got merged: https://github.com/janl/mustache.js/pull/664. This is also what's included in v2.3.1 of mustache.js.

Are you sure you're using the latest version? Or have I misunderstood and you have found another bug?

petrkoutnysw commented:
@phillipj Yeah I have 2.3.1. 

It is quite easy - if you take a look at parse function (line 465), it creates a cache using tags if cache does not exist already.  Then you call render function that calls parse internally without specifying tags. Thus it is getting wrong tokens. Hope it makes sense, I have not been reading mustache code previously.

Anyway, we froze our mustache to 2.3.0.

raymond-lam commented:
@petrkoutnysw Do you have replication steps (the template you used and how you made the call to render it)?

raymond-lam commented:
@petrkoutnysw 

It tried the following code:
```
mustache.parse('<<foo>>bar', ['<<','>>']);
mustache.render('<<foo>>bar', { foo:'biz'})
```

Which outputs "<<foo>>bar" as expected.

Then, if I do:
```
mustache.tags=['<<','>>']
mustache.render('<<foo>>bar', { foo:'biz'});
```

It outputs "bizbar", again as expected.

How are you setting the custom tags (there's only one way as far as I can tell) and calling the `render` function?

petrkoutnysw commented:
@raymond-lam Give me some time please, I am on conference and don't have time to reply right now. Thanks.

phillipj commented:
Maybe these two gives valuable input here? https://github.com/janl/mustache.js/issues/669 & https://github.com/janl/mustache.js/pull/662

raymond-lam commented:
@phillipj @petrkoutnysw Ah yes I see those reported issues now -- I'll have a look.
